akka-actor/src/main/scala/akka/routing/DefaultMetricsBasedResizer.scala

```diff
@@ -0,0 +1,222 @@
+ *
+ */
+@SerialVersionUID(1L)
+case class DefaultMetricsBasedResizer(
+  lowerBound: PoolSize = 1,
+  upperBound: PoolSize = 30,
+  chanceOfScalingDownWhenFull: Double = 0.2,
+  actionInterval: Duration = Duration.ofSeconds(5),
+  numOfAdjacentSizesToConsiderDuringOptimization: Int = 6,
+  exploreStepSize: Double = 0.1,
+  downsizeRatio: Double = 0.8,
+  downsizeAfterUnderutilizedFor: Duration = Duration.ofHours(72),
+  explorationProbability: Double = 0.4,
+  weightOfLatestMetric: Double = 0.5) extends MetricsBasedResizer {
+  /**
+   * Leave package accessible for testing purpose
+   */
+  private[routing] var performanceLog: PerformanceLog = Map.empty
+  /**
+   * Leave package accessible for testing purpose
+   */
+  private[routing] var record: ResizeRecord = ResizeRecord()
+
+  private val random = ThreadLocalRandom.current()
+
+  private def checkParamAsProbability(value: Double, paramName: String): Unit =
+    if (value < 0 || value > 1) throw new IllegalArgumentException(s"$paramName must be between 0 and 1 (inclusive), was: [%s]".format(value))
+
+  private def checkParamAsPositiveNum(value: Double, paramName: String): Unit =
+    if (value < 0) throw new IllegalArgumentException(s"$paramName must be >= 0, was: [%s]".format(value))
+
+  checkParamAsPositiveNum(lowerBound, "lowerBound")
+  checkParamAsPositiveNum(upperBound, "upperBound")
+  if (upperBound < lowerBound) throw new IllegalArgumentException("upperBound must be >= lowerBound, was: [%s] < [%s]".format(upperBound, lowerBound))
+
+  checkParamAsProbability(chanceOfScalingDownWhenFull, "chanceOfScalingDownWhenFull")
+  checkParamAsPositiveNum(numOfAdjacentSizesToConsiderDuringOptimization, "numOfAdjacentSizesToConsiderDuringOptimization")
+  checkParamAsPositiveNum(exploreStepSize, "exploreStepSize")
+  checkParamAsPositiveNum(downsizeRatio, "downsizeRatio")
+  checkParamAsProbability(explorationProbability, "explorationProbability")
+  checkParamAsProbability(weightOfLatestMetric, "weightOfLatestMetric")
+
+  //accessible only for testing purpose
+
+  val historySampleRate: Duration = actionInterval dividedBy 100
+
+  def isTimeForResize(messageCounter: Long): Boolean = {
+    record.checkTime.fold(true)(_.isBefore(LocalDateTime.now.minus(actionInterval)))
+  }
+
+  def reportMessageCount(currentRoutees: immutable.IndexedSeq[Routee], messageCounter: Long): Unit = {
+    val (newPerfLog, newRecord) = updatedStats(currentRoutees, messageCounter)
+
+    performanceLog = newPerfLog
+    record = newRecord
+  }
+
+  private[routing] def updatedStats(currentRoutees: immutable.IndexedSeq[Routee], messageCounter: Long): (PerformanceLog, ResizeRecord) = {
+    val now = LocalDateTime.now
+    val currentSize = currentRoutees.length
+
+    val messagesInRoutees = currentRoutees map {
+      case ActorRefRoutee(a: ActorRefWithCell) ⇒
+        a.underlying match {
+          case cell: ActorCell ⇒
+            cell.mailbox.numberOfMessages + (if (cell.currentMessage != null) 1 else 0)
```

Comment:
I thought about the performance when writing this implementation. Didn't worry too much because this is only called every action period which is usually more than 5 second and its doesn't block anything other than the resizing effort. On the other hand we have to get the mailbox size to collect the metric. Let me know if there is another better way to collect mailbox that I am not aware of.